### ISSUE TITLE
Doesn't save file

### ISSUE THREAD

**Issue by zorque2009** [IpswichMapper/SwiftAddress#6]
The save button wipes the screen but doesn't save any file. At least not in the SwiftAddress folder (there also is no "Keypad Mapper" folder)

**Comment by BTScroggs** [comment]
I am having the same issue.

From logcat output:
```
02-10 18:49:54.682 22283 22283 W addressFile: </osm>
02-10 18:49:54.682 22283 22283 W noteFile:
02-10 18:49:54.683 22283 22283 I ExtStorageAvailable: true
02-10 18:49:54.692  4563 18304 W MediaProvider: Forgot to handle a top level directory in getContentUriForFile?
02-10 18:49:54.699  4563 18304 E MediaProvider: insertFileIfNecessary failed
02-10 18:49:54.699  4563 18304 E MediaProvider: java.lang.IllegalArgumentException: Primary directory SwiftAddress not allowed for content://media/external_primary/file; allowed directories are [Download, Documents]
02-10 18:49:54.699  4563 18304 E MediaProvider:         at com.android.providers.media.MediaProvider.ensureFileColumns(MediaProvider.java:2682)
02-10 18:49:54.699  4563 18304 E MediaProvider:         at com.android.providers.media.MediaProvider.ensureUniqueFileColumns(MediaProvider.java:2347)
02-10 18:49:54.699  4563 18304 E MediaProvider:         at com.android.providers.media.MediaProvider.insertFile(MediaProvider.java:2957)
02-10 18:49:54.699  4563 18304 E MediaProvider:         at com.android.providers.media.MediaProvider.insertInternal(MediaProvider.java:3486)
02-10 18:49:54.699  4563 18304 E MediaProvider:         at com.android.providers.media.MediaProvider.insert(MediaProvider.java:3208)
02-10 18:49:54.699  4563 18304 E MediaProvider:         at com.android.providers.media.MediaProvider.insertFileForFuse(MediaProvider.java:6655)
02-10 18:49:54.699  4563 18304 E MediaProvider:         at com.android.providers.media.MediaProvider.insertFileIfNecessaryForFuse(MediaProvider.java:6742)
02-10 18:49:54.701 22283 22283 W System.err: java.io.IOException: Operation not permitted
02-10 18:49:54.704 22283 22283 W System.err:    at java.io.UnixFileSystem.createFileExclusively0(Native Method)
02-10 18:49:54.704 22283 22283 W System.err:    at java.io.UnixFileSystem.createFileExclusively(UnixFileSystem.java:317)
02-10 18:49:54.704 22283 22283 W System.err:    at java.io.File.createNewFile(File.java:1008)
02-10 18:49:54.704 22283 22283 W System.err:    at layout.StoreHouseNumbers.writeToOsmFile(StoreHouseNumbers.kt:105)
02-10 18:49:54.704 22283 22283 W System.err:    at com.mapitall.SwiftAddress.MainActivity.saveFile(MainActivity.kt:478)
02-10 18:49:54.704 22283 22283 W System.err:    at java.lang.reflect.Method.invoke(Native Method)
02-10 18:49:54.704 22283 22283 W System.err:    at androidx.appcompat.app.AppCompatViewInflater$DeclaredOnClickListener.onClick(AppCompatViewInflater.java:409)
02-10 18:49:54.704 22283 22283 W System.err:    at android.view.View.performClick(View.java:7448)
02-10 18:49:54.704 22283 22283 W System.err:    at android.view.View.performClickInternal(View.java:7425)
02-10 18:49:54.704 22283 22283 W System.err:    at android.view.View.access$3600(View.java:810)
02-10 18:49:54.704 22283 22283 W System.err:    at android.view.View$PerformClick.run(View.java:28305)
02-10 18:49:54.704 22283 22283 W System.err:    at android.os.Handler.handleCallback(Handler.java:938)
02-10 18:49:54.704 22283 22283 W System.err:    at android.os.Handler.dispatchMessage(Handler.java:99)
02-10 18:49:54.704 22283 22283 W System.err:    at android.os.Looper.loop(Looper.java:223)
02-10 18:49:54.704 22283 22283 W System.err:    at android.app.ActivityThread.main(ActivityThread.java:7660)
02-10 18:49:54.704 22283 22283 W System.err:    at java.lang.reflect.Method.invoke(Native Method)
02-10 18:49:54.704 22283 22283 W System.err:    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
02-10 18:49:54.704 22283 22283 W System.err:    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
```

**Comment by IpswichMapper** [comment]
@BTScroggs Hmm, interesting. What version of android are you using?

**Comment by BTScroggs** [comment]
Android 11, I think applications have to ask for a folder rather than create one directly.

Also as more of a feature request, it would be nice if it didn't delete the data if the save was unsuccessful.

**Comment by IpswichMapper** [comment]
Yup, android 11 forces app to use "scoped storage", meaning trying to create a SwiftAddress directory in the root folder is not allowed. (You can only save in certain directories, for example images can be saved in Pictures or DCIM).

I've just reworked the entire save feature so that it uses your files app to choose a location to save.

A zip file containing all the images and addresses you have taken is then stored in that location.

The advantage of this is that the app no longer needs access to storage :)

-- 
11 Feb 2021, 16:14 by notifications@github.com:

**Comment by IpswichMapper** [comment]
Also the whole save file operation is put into a try{} catch{} block, and the deleting of data is done after everything has saved. So now, if the data fails to save, an error will occur in the try{} block which will prevent further code from being executed.

This should save the data if an error occurs when the app tries to save data.



--  

10 Feb 2021, 18:30 by notifications@github.com:

**Comment by IpswichMapper** [comment]
@BTScroggs  how did you get that logcat? You cannot see the package name there, so how did you know how to get the logs to pick out?

**Comment by BTScroggs** [comment]
Great work on fixing it so quick looking forward to the next release.

Only ran logcat for a short time while pressing the save button. I could see there was a couple of debug messages about the files. After a few more goes while looking at your code got lucky with the error.

**Comment by IpswichMapper** [comment]
new version released with fix just now (version 0.2)
